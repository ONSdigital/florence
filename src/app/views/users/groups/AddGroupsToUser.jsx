import React, { useState, useEffect, useCallback } from "react";
import filter from "lodash/filter";
import { withRouter } from "react-router";
import PropTypes from "prop-types";
import notifications from "../../../utilities/notifications";
import { useInput } from "../../../hooks/useInput";
import BackButton from "../../../components/back-button";
import FormFooter from "../../../components/form-footer";
import GroupsTable from "../../../components/table";
import Loader from "../../../components/loader/Loader";
import Magnifier from "../../../icons/Magnifier";
import User from "./User";

const notification = {
    type: "warning",
    message: "User is already a member of this group.",
    autoDismiss: 5000,
};

function AddGroupsToUser(props) {
    const id = props.params.id;

    useEffect(() => {
        if (id) {
            loadUser(id);
            loadGroups();
        }
    }, [id]);

    const { loading, user, groups, loadUser, loadingGroups, loadGroups, addGroupsToUser, isAdding, rootPath } = props;
    const [isSubmitting, setIsSubmitting] = useState(isAdding);
    const [search, setSearch] = useInput("");
    const [userGroups, setUserGroups] = useState([]);

    const hasNewValues = userGroups.length > 0;
    const routerWillLeave = nextLocation => {
        if (hasNewValues && !isSubmitting) return "Your work is not saved! Are you sure you want to leave?";
    };

    useEffect(() => {
        props.router.setRouteLeaveHook(props.route, routerWillLeave);
    });

    const handleRemove = id => setUserGroups(prevState => prevState.filter(group => group.id !== id));

    const handleAdd = group => {
        if (userGroups.includes(group)) {
            notifications.add(notification);
            return;
        }
        setIsSubmitting(true);
        setUserGroups(prevState => prevState.concat(group));
    };

    const getFilteredGroups = useCallback(() => {
        return filter(groups, group => group.name?.toLowerCase().includes(search.value.toLowerCase()));
    }, [search.value]);

    return (
        <div className="grid grid--justify-space-around form-with-sticky-footer">
            <div className="grid__col-11 grid__col-md-9">
                <BackButton redirectUrl={`${rootPath}/users`} classNames="margin-top--2" />
                <div className="grid grid--justify-space-between">
                    <div className="grid__col-md-5">
                        {loading ? (
                            <Loader classNames="grid grid--align-center grid--align-self-center" />
                        ) : (
                            <User {...user} userGroups={userGroups} handleRemove={handleRemove} />
                        )}
                    </div>
                    <div className="grid__col-md-6">
                        <h2 className="margin-top--1 margin-bottom--1">Add a team for the user to join</h2>
                        <div className="search__input-group margin-bottom--1">
                            <Magnifier classes="search__icon-magnifier" viewBox="0 0 28 28" />
                            <label htmlFor="search" className="visually-hidden">
                                Search teams by name
                            </label>
                            <input role="search" name="search" placeholder="Search teams by name" {...search} />
                        </div>
                        {loadingGroups ? (
                            <Loader classNames="grid grid--align-center grid--align-self-center" />
                        ) : (
                            <GroupsTable testid="groups-table" items={search.value ? getFilteredGroups() : groups} handleClick={handleAdd} />
                        )}
                    </div>
                </div>
            </div>
            <FormFooter
                hasNewValues={hasNewValues}
                loading={isAdding}
                redirectUrl={`${rootPath}/users`}
                handleSubmit={() =>
                    addGroupsToUser(
                        id,
                        userGroups.map(group => group.id)
                    )
                }
            />
        </div>
    );
}

AddGroupsToUser.propTypes = {
    user: PropTypes.object,
    groups: PropTypes.array,
    isAdding: PropTypes.bool,
    loading: PropTypes.bool,
    params: PropTypes.object,
    addGroupsToUser: PropTypes.func.isRequired,
    loadGroups: PropTypes.func.isRequired,
    loadUser: PropTypes.func.isRequired,
};

export default withRouter(AddGroupsToUser);
