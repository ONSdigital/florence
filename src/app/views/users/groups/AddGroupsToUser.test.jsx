import React from "react";
import renderer from "react-test-renderer";
import { render, screen, within, getByTestId } from "../../../utilities/tests/test-utils";
import AddGroupsToUser from "./AddGroupsToUser";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";
import { groups, user } from "../../../utilities/tests/mockData";

const defaultProps = {
    user: null,
    groups: [],
    isAdding: false,
    loading: false,
    addGroupsToUser: jest.fn(),
    loadUser: jest.fn(),
    loadGroups: jest.fn(),
    router: { setRouteLeaveHook: jest.fn() },
};
const props = {
    ...defaultProps,
    user: user,
    groups: groups,
};
const setRouteLeaveHook = jest.fn();

describe("AddGroupsToUser", () => {
    it("matches the snapshot", () => {
        const wrapper = renderer.create(<AddGroupsToUser.WrappedComponent {...defaultProps} params={{ router: setRouteLeaveHook }} />);
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it("requests user data on component load", () => {
        render(<AddGroupsToUser.WrappedComponent {...defaultProps} params={{ userID: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);
        expect(defaultProps.loadUser).toHaveBeenCalledWith("test.user-1498@ons.gov.uk");
    });

    it("requests groups on component load", () => {
        render(<AddGroupsToUser.WrappedComponent {...defaultProps} params={{ userID: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);
        expect(defaultProps.loadGroups).toHaveBeenCalled();
    });

    it("shows Back Button", () => {
        render(<AddGroupsToUser.WrappedComponent {...defaultProps} params={{ userID: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);
        expect(screen.getByText(/Back/i)).toBeInTheDocument();
    });

    it("shows Groups heard, search and message if no groups", () => {
        render(<AddGroupsToUser.WrappedComponent {...defaultProps} params={{ userID: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);
        expect(screen.getByText(/add a team for the user to join/i)).toBeInTheDocument();
        expect(screen.getByPlaceholderText("Search teams by name")).toHaveValue("");
    });

    it("shows Footer Section with Buttons", () => {
        render(<AddGroupsToUser.WrappedComponent {...defaultProps} params={{ userID: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);
        expect(screen.getByTestId("form-footer")).toBeInTheDocument();
        expect(screen.getByText(/Save changes/i)).toBeInTheDocument();
        expect(screen.getByText(/Cancel/i)).toBeInTheDocument();
    });

    describe("when loading data", () => {
        it("shows spinner", () => {
            const props = { ...defaultProps, loading: true };
            render(<AddGroupsToUser.WrappedComponent {...props} params={{ userID: "", router: setRouteLeaveHook }} />);

            expect(screen.getByTestId("loader")).toBeInTheDocument();
        });
    });

    describe("where user and groups fetched", () => {
        it("shows user details", () => {
            render(<AddGroupsToUser.WrappedComponent {...props} params={{ userID: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);
            expect(screen.getByText(/Back/i)).toBeInTheDocument();
            expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent(/test user-1498/i);
            expect(screen.getByText("test.user-1498@ons.gov.uk")).toBeInTheDocument();
            expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent(/test user-1498/i);
            expect(within(screen.getByTestId("user")).getByRole("heading", { level: 2 })).toHaveTextContent(/Team Member/i);
        });

        it("shows groups", () => {
            render(<AddGroupsToUser.WrappedComponent {...props} params={{ userID: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);
            const GroupsSection = screen.getByTestId("groups-table");

            expect(screen.getByText(/add a team for the user to join/i)).toBeInTheDocument();
            expect(screen.getByPlaceholderText("Search teams by name")).toHaveValue("");
            expect(screen.getByText(/my test group/i)).toBeInTheDocument();
            expect(screen.getByText(/my test group/i)).toBeInTheDocument();
            expect(screen.getAllByRole("button", { name: "Add" })).toHaveLength(3);
        });

        it("adds user to groups", () => {
            render(<AddGroupsToUser.WrappedComponent {...props} params={{ userID: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);
            const GroupsSection = screen.getByTestId("groups-table");

            expect(screen.getByRole("heading", { level: 2, name: "Add a team for the user to join" })).toBeInTheDocument();
            expect(screen.getByPlaceholderText("Search teams by name")).toHaveValue("");
            expect(screen.getByText(/admins/i)).toBeInTheDocument();
            expect(screen.getByText(/my first group/i)).toBeInTheDocument();
            const add_buttons = screen.getAllByRole("button", { name: "Add" });

            userEvent.click(add_buttons[0]);

            expect(screen.getByTestId("form-footer")).toBeInTheDocument();
            expect(within(screen.getByTestId("user")).getByText("my test group")).toBeInTheDocument();
            expect(within(screen.getByTestId("form-footer")).getByText("You have unsaved changes")).toBeInTheDocument();

            userEvent.click(screen.getByText(/save changes/i));
            expect(props.addGroupsToUser).toHaveBeenCalledWith("test.user-1498@ons.gov.uk", ["my test group"]);
        });
    });
});
