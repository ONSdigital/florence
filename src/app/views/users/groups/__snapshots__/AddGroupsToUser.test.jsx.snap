// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddGroupsToUser matches the snapshot 1`] = `
<div
  className="grid grid--justify-space-around form-with-sticky-footer"
>
  <div
    className="grid__col-11 grid__col-md-9"
  >
    <div
      className="margin-top--2"
    >
      ◀ 
      <a
        onClick={[Function]}
        role="link"
        style={Object {}}
      >
        Back
      </a>
    </div>
    <div
      className="grid grid--justify-space-between"
    >
      <div
        className="grid__col-md-5"
      >
        <div
          data-testid="user"
        >
          <h1
            className="margin-top--1 margin-bottom--1"
          >
            undefined undefined
          </h1>
          <p />
          <h2
            className="margin-top--1"
          >
            Team Member
          </h2>
          <p>
            User is not a member of a team. Please add them to a team.
          </p>
        </div>
      </div>
      <div
        className="grid__col-md-6"
      >
        <h2
          className="margin-top--1 margin-bottom--1"
        >
          Add a team for the user to join
        </h2>
        <div
          className="search__input-group margin-bottom--1"
        >
          <svg
            className="search__icon-magnifier"
            role="presentation"
            viewBox="0 0 28 28"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m20.011 16.352h-1.2647l-0.4482-0.4014c1.5689-1.6947 2.5134-3.8948 2.5134-6.2881 0-5.3368-4.6587-9.6627-10.406-9.6627-5.7472 0-10.406 4.3259-10.406 9.6627 0 5.3367 4.6587 9.6626 10.406 9.6626 2.5775 0 4.9469-0.877 6.7719-2.3339l0.4323 0.4163v1.1743l8.0045 7.418 2.3854-2.215-7.9886-7.4328zm-9.6055 0c-3.9862 0-7.2041-2.988-7.2041-6.6895s3.2178-6.6895 7.2041-6.6895c3.9863 0 7.2042 2.988 7.2042 6.6895s-3.2179 6.6895-7.2042 6.6895z"
              fill="#808080"
            />
          </svg>
          <label
            className="visually-hidden"
            htmlFor="search"
          >
            Search teams by name
          </label>
          <input
            name="search"
            onChange={[Function]}
            placeholder="Search teams by name"
            role="search"
            value=""
          />
        </div>
        <p>
          Nothing to show.
        </p>
      </div>
    </div>
  </div>
  <div
    className="grid grid--justify-space-around padding-bottom--1 padding-top--1 form-footer"
    data-testid="form-footer"
  >
    <div
      className="grid__col-9"
    >
      <div
        className="grid grid--align-baseline"
      >
        <button
          className="btn btn--positive btn--margin-right"
          onClick={[Function]}
          type="submit"
        >
          Save changes
        </button>
        <a
          className="btn btn--invert-primary btn--margin-right"
          onClick={[Function]}
          role="button"
          style={Object {}}
        >
          Cancel
        </a>
      </div>
    </div>
  </div>
</div>
`;
